File: frontend/js/collections.js
/**
 * Collections mixin — dry-run diffs, syncing, history, posters.
 */
import { api } from './api.js';

export function collectionsMixin() {
  return {
    async loadCollections() {
      try {
        const res = await api.collections.list();
        this.collections = res.collections;
        this.dryRunDefault = res.dry_run_default;
      } catch (e) {
        this.error = e.message;
      }
    },

    /** Preview. Never writes, and ignores the guards so you can always see
     *  what a refused sync *would* have done. */
    async showDiff(ruleId) {
      this.diffModal = { open: true, loading: true, diff: null, ruleId, guarded: false };
      try {
        this.diffModal = {
          open: true, loading: false, guarded: false, ruleId,
          diff: await api.collections.diff(ruleId),
        };
      } catch (e) {
        this.diffModal = { open: false, loading: false, diff: null, ruleId, guarded: false };
        this.showToast(e.message, false);
      }
    },

    closeDiff() {
      this.diffModal = { open: false, loading: false, diff: null, ruleId: null, guarded: false };
    },

    /**
     * Apply a rule. A 409 means a guard refused it — the response carries the
     * diff, so we show exactly what was refused and offer an explicit override
     * rather than a bare error.
     */
    async syncCollection(ruleId, { dryRun = null, force = false } = {}) {
      this.syncing = ruleId;
      try {
        const diff = await api.collections.sync(ruleId, { dryRun, force });
        this.diffModal = { open: true, loading: false, diff, ruleId, guarded: false };
        if (diff.applied) {
          this.showToast(`Synced: +${diff.add_count} / −${diff.remove_count}`);
          await this.loadCollections();
        } else {
          this.showToast('Dry run — nothing was written to Plex.');
        }
      } catch (e) {
        if (e.status === 409 && e.payload?.diff) {
          this.diffModal = {
            open: true, loading: false, ruleId,
            diff: e.payload.diff, guarded: true, guardMessage: e.message,
          };
        } else {
          this.showToast(e.message, false);
        }
      } finally {
        this.syncing = null;
      }
    },

    /**
     * Apply from the diff modal. Confirms first when dry run is on, because the
     * global switch says nothing gets written and this deliberately ignores it —
     * silently overriding a safety setting the user turned on is not on.
     */
    async applyFromDiff() {
      const diff = this.diffModal.diff;
      if (this.dryRunDefault) {
        const ok = await this.showConfirm({
          title: 'Write to Plex now?',
          message:
            `Dry run is on, so nothing has been written so far. This applies ` +
            `${diff.add_count} label${diff.add_count === 1 ? '' : 's'}` +
            (diff.remove_count ? ` and removes ${diff.remove_count}` : '') +
            ` in Plex right now, and creates the collection.\n\n` +
            `Reversible with "unsync" on the collection card.`,
          confirmLabel: 'Write to Plex',
          danger: true,
        });
        if (!ok) return;
      }
      await this.syncCollection(this.diffModal.ruleId, { dryRun: false });
    },

    async syncAll() {
      const ok = await this.showConfirm({
        title: 'Sync every enabled rule',
        message: this.dryRunDefault
          ? 'Dry run is on, so this only reports what would change.'
          : 'This will write labels to Plex for every enabled rule.',
        confirmLabel: 'Sync all',
        danger: !this.dryRunDefault,
      });
      if (!ok) return;
      this.syncing = 'all';
      try {
        const res = await api.collections.syncAll();
        const failed = res.results.filter((r) => !r.ok).length;
        // A guard refusal is the app being cautious, not a failure — amber.
        this.showToast(
          `${res.count - failed} of ${res.count} rules synced${failed ? `, ${failed} refused` : ''}.`,
          failed ? 'warn' : true,
        );
        await this.loadCollections();
      } catch (e) {
        this.showToast(e.message, false);
      } finally {
        this.syncing = null;
      }
    },

    async unsyncCollection(ruleId, name, stripAll = false) {
      const ok = await this.showConfirm({
        title: stripAll ? 'Force-remove from Plex' : 'Remove from Plex',
        message: stripAll
          ? `Strip the label from EVERY item in Plex carrying it — including any ` +
            `applied by hand — and delete the "${name}" collection.\n\n` +
            `Your :pin and :veto labels are still left alone.`
          : `Strip the Plexlection label from every item we applied it to and delete ` +
            `the "${name}" collection.\n\nYour own :pin and :veto labels are left alone.`,
        confirmLabel: stripAll ? 'Force remove' : 'Remove',
        danger: true,
      });
      if (!ok) return;
      try {
        const res = await api.collections.unsync(ruleId, stripAll);
        // Leftovers after a normal unsync are hand-applied labels — say so and
        // point at the force option rather than leaving them a mystery.
        if (res.still_labelled) {
          this.showToast(
            `Removed ${res.removed}. ${res.still_labelled} labels remain that we didn't apply — ` +
            `use force-remove to clear those too.`, 'warn');
        } else {
          this.showToast(`Removed ${res.removed} labels; collection deleted.`);
        }
        await this.loadCollections();
      } catch (e) {
        this.showToast(e.message, false);
      }
    },

    async loadHistory(ruleId) {
      try {
        const res = await api.collections.history(ruleId);
        this.syncHistory = res.history;
        this.historyFor = ruleId;
      } catch {
        this.syncHistory = [];
      }
    },

    // ── presentation details (title, summary, order) ────────────────────
    openDetails(c) {
      if (this.detailsFor === c.id) {
        this.detailsFor = null;
        return;
      }
      this.detailsFor = c.id;
      this.detailsDraft = {
        collection_title: c.collection_title ?? '',
        collection_sort_title: c.collection_sort_title ?? '',
        collection_summary: c.collection_summary ?? '',
        collection_sort: c.collection_sort ?? '',
      };
    },

    async saveDetails(ruleId) {
      try {
        await api.rules.update(ruleId, { ...this.detailsDraft });
        this.detailsFor = null;
        await this.loadCollections();
        this.showToast('Details saved — they apply on the next sync.');
      } catch (e) {
        this.showToast(e.message, false);
      }
    },

    async uploadPoster(ruleId, event) {
      const file = event.target.files?.[0];
      if (!file) return;
      try {
        await api.collections.uploadPoster(ruleId, file);
        this.showToast('Poster saved — it uploads on the next sync.');
        await this.loadCollections();
      } catch (e) {
        this.showToast(e.message, false);
      } finally {
        event.target.value = '';
      }
    },

    async removePoster(ruleId) {
      try {
        await api.collections.removePoster(ruleId);
        await this.loadCollections();
      } catch (e) {
        this.showToast(e.message, false);
      }
    },
  };
}

File: backend/routers/collections.py
"""Collection sync: dry-run diffs, applying, history, posters."""
import hashlib
import json
import time
from pathlib import Path

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile
from fastapi.responses import FileResponse, Response

from backend import startup
from backend.common.auth import require_auth
from backend.common.config import config
from backend.common.logging_config import get_logger
from backend.routers.rules import _row_to_rule

logger = get_logger(__name__)
router = APIRouter(prefix="/api/collections", tags=["collections"])

MAX_POSTER_BYTES = 16 * 1024 * 1024


async def _get_rule(rule_id: int) -> dict:
    row = await startup.db.fetch_one("SELECT * FROM rules WHERE id = ?", (rule_id,))
    if row is None:
        raise HTTPException(status_code=404, detail="Rule not found")
    return _row_to_rule(row)


@router.get("", dependencies=[Depends(require_auth)])
async def list_collections():
    """Every rule with its sync state, for the Collections tab."""
    rows = await startup.db.fetch_all(
        "SELECT r.*, "
        "  (SELECT COUNT(*) FROM sync_membership m WHERE m.rule_id = r.id) AS synced_count "
        "FROM rules r ORDER BY r.name COLLATE NOCASE"
    )
    out = []
    for row in rows:
        rule = _row_to_rule(row)
        label, pin, veto = startup.sync_engine.labels_for(rule["slug"])
        rule["labels"] = {"main": label, "pin": pin, "veto": veto}
        out.append(rule)
    return {
        "collections": out,
        "count": len(out),
        "dry_run_default": startup.settings_store.get().safety.dry_run,
    }


@router.post("/{rule_id}/diff", dependencies=[Depends(require_auth)])
async def diff_collection(rule_id: int):
    """What a sync would change. Never writes to Plex."""
    rule = await _get_rule(rule_id)
    diff = await startup.sync_engine.sync_rule(rule, dry_run=True, force=True)
    return diff.as_dict()


@router.post("/{rule_id}/sync", dependencies=[Depends(require_auth)])
async def sync_collection(rule_id: int, dry_run: bool | None = None, force: bool = False):
    """Apply a rule to Plex.

    `dry_run` defaults to the global safety setting, which starts on — nothing
    reaches Plex until it's deliberately turned off. `force` overrides the
    guards and is what the UI's second-click "sync anyway" sends.
    """
    rule = await _get_rule(rule_id)
    if dry_run is None:
        dry_run = startup.settings_store.get().safety.dry_run

    # SyncGuardError -> 409 with the diff attached (see main.py's handler),
    # so the UI can show exactly what it refused and offer the override.
    diff = await startup.sync_engine.sync_rule(
        rule, dry_run=dry_run, force=force, trigger="manual"
    )

    if startup.broadcaster and diff.applied:
        startup.broadcaster.emit("sync_done", {
            "rule_id": rule_id, "rule_name": rule["name"],
            "added": len(diff.add), "removed": len(diff.remove),
        })
    return diff.as_dict()


@router.post("/sync-all", dependencies=[Depends(require_auth)])
async def sync_all(dry_run: bool | None = None):
    rows = await startup.db.fetch_all("SELECT * FROM rules WHERE enabled = 1")
    if dry_run is None:
        dry_run = startup.settings_store.get().safety.dry_run

    results = []
    for row in rows:
        rule = _row_to_rule(row)
        try:
            diff = await startup.sync_engine.sync_rule(
                rule, dry_run=dry_run, trigger="manual"
            )
            results.append({"rule": rule["name"], "ok": True, **diff.as_dict()})
        except Exception as exc:
            # One guarded or failing rule must not stop the rest.
            results.append({"rule": rule["name"], "ok": False, "detail": str(exc)})
    return {"results": results, "count": len(results), "dry_run": dry_run}


@router.post("/{rule_id}/unsync", dependencies=[Depends(require_auth)])
async def unsync_collection(rule_id: int, strip_all: bool = False):
    """Remove our labels and delete the collection. Leaves :pin/:veto alone.

    `strip_all=true` removes the label from every item in Plex carrying it,
    not just the ones we recorded applying — the reset for when our records and
    Plex have diverged.
    """
    rule = await _get_rule(rule_id)
    result = await startup.sync_engine.unsync_rule(rule, strip_all=strip_all)
    return {"success": True, **result}


@router.get("/{rule_id}/history", dependencies=[Depends(require_auth)])
async def collection_history(rule_id: int, limit: int = 20):
    rows = await startup.db.fetch_all(
        "SELECT * FROM sync_history WHERE rule_id = ? ORDER BY started_at DESC LIMIT ?",
        (rule_id, limit),
    )
    history = []
    for row in rows:
        entry = dict(row)
        try:
            entry["detail"] = json.loads(entry.pop("detail_json") or "{}")
        except json.JSONDecodeError:
            entry["detail"] = {}
        history.append(entry)
    return {"history": history, "count": len(history)}


@router.get("/{rule_id}/poster", dependencies=[Depends(require_auth)])
async def get_poster(rule_id: int):
    """The card image: the uploaded poster when one exists, else Plex's own
    art for the collection — Plex composites one automatically, so every
    synced collection has something to show."""
    rule = await _get_rule(rule_id)

    ref = rule.get("poster_ref")
    if ref and Path(ref).exists():
        media = "image/jpeg" if ref.endswith(".jpg") else "image/png"
        # The frontend cache-busts on poster_fp, so long caching is safe.
        return FileResponse(ref, media_type=media,
                            headers={"Cache-Control": "private, max-age=86400"})

    if rule.get("collection_rating_key"):
        data = await startup.get_plex().collection_poster(rule["collection_rating_key"])
        if data:
            return Response(content=data, media_type="image/jpeg",
                            headers={"Cache-Control": "private, max-age=3600"})

    raise HTTPException(status_code=404, detail="No poster available.")


@router.post("/{rule_id}/poster", dependencies=[Depends(require_auth)])
async def upload_poster(rule_id: int, file: UploadFile = File(...)):
    """Store a poster for the collection.

    Hashed on save so a nightly sync doesn't re-upload an unchanged image —
    every Plex write bumps updatedAt and cascades into client refreshes.
    """
    rule = await _get_rule(rule_id)
    data = await file.read()
    if len(data) > MAX_POSTER_BYTES:
        raise HTTPException(status_code=413, detail="Poster must be 16MB or smaller.")
    if not (file.content_type or "").startswith("image/"):
        raise HTTPException(status_code=415, detail="That doesn't look like an image.")

    config.poster_dir().mkdir(parents=True, exist_ok=True)
    suffix = ".jpg" if "jpeg" in (file.content_type or "") else ".png"
    path = config.poster_dir() / f"{rule['slug']}{suffix}"
    path.write_bytes(data)

    await startup.db.execute(
        "UPDATE rules SET poster_ref = ?, poster_fp = ?, updated_at = ? WHERE id = ?",
        (str(path), hashlib.sha1(data).hexdigest(), int(time.time()), rule_id),
    )
    return {"success": True, "poster_ref": str(path), "bytes": len(data)}


@router.delete("/{rule_id}/poster", dependencies=[Depends(require_auth)])
async def remove_poster(rule_id: int):
    rule = await _get_rule(rule_id)
    if rule.get("poster_ref"):
        Path(rule["poster_ref"]).unlink(missing_ok=True)
    await startup.db.execute(
        "UPDATE rules SET poster_ref = NULL, poster_fp = NULL WHERE id = ?", (rule_id,)
    )
    return {"success": True}
